Fix serialisation of Bimap

nc6 opened the pull request:
This PR addresses a number of issues related to Bimap serialisation:
- The existing serialisation format involved encoding both the forward
and reverse maps, leaving our on-disk state twice as large as it needed
to be.
- A change in #1886 resulted in the internal structure of Bimap
changing, and a consequent inadvertant change in the serialised format.
We want to avoid this wherever possible, since this entails all nodes
needing to rebuild the Ledger DB from scratch.

Consequently, we make the following changes:
- Newly serialised 'Bimap's will serialise only the forward map. This is
then decoded to populate both directions.
- The decoder will also accept the previously encoded format, except
that in this case it ignores the backwards encoding. This should allow
it to decode 'Bimap's encoded before #1886.

The current implementation of 'biMapFromList' does not look terribly
efficient; this commit does not (yet) address that.

nc6 committed:
Fix serialisation of Bimap

This PR addresses a number of issues related to Bimap serialisation:
- The existing serialisation format involved encoding both the forward
and reverse maps, leaving our on-disk state twice as large as it needed
to be.
- A change in #1886 resulted in the internal structure of Bimap
changing, and a consequent inadvertant change in the serialised format.
We want to avoid this wherever possible, since this entails all nodes
needing to rebuild the Ledger DB from scratch.

Consequently, we make the following changes:
- Newly serialised 'Bimap's will serialise only the forward map. This is
then decoded to populate both directions.
- The decoder will also accept the previously encoded format, except
that in this case it ignores the backwards encoding. This should allow
it to decode 'Bimap's encoded before #1886.

The current implementation of 'biMapFromList' does not look terribly
efficient; this commit does not (yet) address that.

mrBliss commented:
It would be good to have a golden test that covers the `_ptrs :: !(Bimap Ptr (Credential 'Staking era))` of `DState`.

nc6 committed:
Add faster means for deserialising 'Bimap'.

This gives around a 2x speedup. There would be a way to go faster, but
it's tricky without digging into the internals of the Map
implementation.

Also add a benchmark to compare these.

mrBliss commented on semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs:
```diff
@@ -229,6 +230,12 @@ biMapFromList comb xs = foldr addEntry biMapEmpty xs
+biMapFromAscDistinctList xs = MkBiMap bmForward bmBackward
+  where
+    bmForward = Map.fromDistinctAscList xs
+    bmBackward = foldr (uncurry $ flip addBack) Map.empty xs
```
I'm wondering how this compares to `foldl' (uncurry addBack) Map.empty xs`.

TimSheard replied:
I think it is worth adding to the benchmark a try of the foldl'  way. Being strict and a left fold should make things faster.

mrBliss commented on semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs:
```diff
@@ -137,22 +149,43 @@ data BiMap v a b where MkBiMap:: (v ~ b) => !(Map.Map a b) -> !(Map.Map b (Set.S
-    !y <- fromCBOR
-    return (MkBiMap x y)
+instance (Ord a, Ord b, ToCBOR a, ToCBOR b) => ToCBOR (BiMap b a b) where
+  -- The `toCBOR` instance encodes only the forward map. We wrap this in a
```
```suggestion
  -- The 'toCBOR' instance encodes only the forward map. We wrap this in a
```

mrBliss commented on semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs:
```diff
@@ -137,22 +149,43 @@ data BiMap v a b where MkBiMap:: (v ~ b) => !(Map.Map a b) -> !(Map.Map b (Set.S
+      -- further, the reverse encoding was from 'b' to 'a', not the current 'b'
+      -- to 'Set a', and hence the dropper reflects that.
+      2 -> do
+        !x <- decodeMapAsBimap
+        dropMap (void $ fromCBOR @b) (void $ fromCBOR @a)
+        return x
```
I assume you have verified that the backwards compatibility indeed works as expected? :slightly_smiling_face:

TimSheard replied:
There are some tests that test that the forward and backward map encode the same information.

nc6 replied:
Trying to do that, which is why this is still draft!

nc6 replied:
I've confirmed that this is able to decode the old snapshot.

mrBliss commented on semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs:
```diff
@@ -137,22 +149,43 @@ data BiMap v a b where MkBiMap:: (v ~ b) => !(Map.Map a b) -> !(Map.Map b (Set.S
+  -- The `toCBOR` instance encodes only the forward map. We wrap this in a
+  -- length-one list because a _previous_ encoding wrote out both maps, and we
+  -- can easily use the list length token to distinguish between them.
+  toCBOR (MkBiMap l _) = encodeListLen 1 <> toCBOR l
```
We had to do a similar thing in consensus once. I removed that extra `encodeListLen 1` the next time the serialisation had to change. We can keep this in mind the next time the ledger format has to change.

TimSheard reviewed (approved):
Everything looks good. I would suggest, adding a foldl' version of decodeMapAsBimap to the benchmark, running it at various lengths, and making an informed decision of which one to choose.
